README
			Strabon
		      -----------

Introduction
============
Strabon is a fully implemented semantic geospatial database system that can be
used to store linked geospatial data expressed in RDF and query them using an
extension of SPARQL. Strabon supports spatial selections, spatial joins, a rich
set of spatial functions similar to those offered by geospatial relational
database systems, support for multiple Coordinate Reference Systems and widely
used serializations for geometric objects such as WKT and GML. Strabon is built
on top of the well-known RDF store Sesame and extends Sesame’s components to be
able to manage thematic and spatial data that are stored in PostGIS.

The development of Strabon started in the context of European FP7 project
SemsorGrid4Env (Semantic Sensor Grids for Rapid Application Development for 
Environmental Management) [http://www.semsorgrid4env.eu/]. Starting September
2011, Strabon is being utilized and extended with new functionalities in the
FP7 project TELEIOS (Virtual Observatory Infrastructure for Earth Observation
Data) [http://www.earthobservatory.eu/] which our group leads.

The query language of Strabon is called stSPARQL. stSPARQL can be used to query
data represented in an extension of RDF called stRDF. stRDF and stSPARQL have
been designed for representing and querying geospatial data that changes over
time (e.g., the growth of a city over the years due to new developments). 

Given the very close relationship between stSPARQL and GeoSPARQL which is a 
recent OGC standard for an extension of SPARQL for querying geospatial metadata,
we recently provided support for the Core, Geometry and Geometry Topology 
extension of GeoSPARQL. 


Conformance to GeoSPARQL
========================
Strabon implements the `Core', the `Topology Vocabulary Extension', the
`Geometry Extension', the `Geometry Topology Extension', and the `RDFS Entailment
Extension' except for Req. 25
(http://www.opengis.net/spec/geosparql/1.0/req/rdfs-entailment-extension/bgp-rdfs-ent).

With respect to GML, Strabon supports the GML Profile corresponding to Simple
Features, that is, GML Simple Features Profile 2.0. 


Strabon Homepage
================
The homepage of Strabon is at http://www.strabon.di.uoa.gr/.


Demo
====
You can find a demo of the system Strabon at http://test.strabon.di.uoa.gr/NOA/.


How to build and run Strabon from command line
==============================================
Assuming you have already downloaded Strabon and you are in the top-level
directory of Strabon, issue the following command to build it from command line:

	$ mvn clean package

In order to run automatically the JUnit tests, pass the option `-DskipTests=false' 
to the above command. The complete command is the following:

	$ mvn -DskipTests=false clean package

Alternatively, it is possible to run a specific test. Supposing that the name of
the corresponding class is `TestName', then you can run only this
test using the following command:

	$ mvn test -DfailIfNoTests=false -DskipTests=false -Dtest=TestName

In case of an error during building of Strabon and assuming that the error does
not come from the JUnit tests, please have a look at the `Known Issues' section
below. If none of the known issues of that section applies, please contact the
developers through the Strabon mailing-list or submit a corresponding bug
(see `Bugs' section below).

After you have successfully built Strabon, you have access to the following
components:

  * Strabon Endpoint

    This is a SPARQL endpoint for Strabon. It is distributed as a war file so
    you may deploy it in a Tomcat container. You may find the war file under
    directory `endpoint/target'.

  * Strabon Endpoint (standalone)

    This is a SPARQL endpoint for Strabon like the above one, but it differs
    only in that it does not require the user to have already set up a Tomcat
    container. The standalone Strabon Endpoint may be run by issuing the
    following command:

    	$ java -jar endpoint-exec/target/strabon-endpoint-executable-${version}.jar

    After issuing the above command, you may access the Strabon Endpoint at
    the following URL: <http://localhost:8080/>.

    Please see the page at <http://hg.strabon.di.uoa.gr/Strabon/rev/674f8f91162b>
    to find out other options that you may pass to the Tomcat container that
    will run by the above command.

    SPECIAL NOTE: if you need to configure the connection details to the
    underlying database, you may do so in two ways:
      1. By modifying file `endpoint/WebContent/WEB-INF/connection.properties'
         before building Strabon and executing the above command.
      2. After executing the above command, by visiting the following page by a
         browser: <http://localhost:8080/ChangeConnection>

  * Strabon Endpoint Client

    This is a Java client for interacting with Strabon Endpoint or any other
    SPARQL endpoint. It is packaged as a jar file and may be found under
    directory `endpoint-client/target/' with name
    `strabon-endpoint-client-${version}.jar'. This jar contains any dependencies
    to other code, so may copy and paste it to your project and start playing
    with the code immediately.

  * Strabon script

    The `strabon' script is located under the `scripts/' directory and it is the
    main command-line tool for interacting with Strabon. You may use it to store
    RDF data with geospatial information or query/update it using one of
    stSPARQL or GeoSPARQL query languages.

  * Endpoint script

    The `endpoint' script is located under the `scripts' directory and it is the
    main command-line tool for interacting with a `Strabon Endpoint'. You may
    use it to do any operation you would like to do with the `strabon' script
    above, but in contrast to the `strabon' script you need to have access to a
    Strabon endpoint. Of course, the `Strabon Endpoint Client' component above
    can be used as well as a command-line tool. At the time of writing, the
    `Strabon Endpoint Client' component supports only querying of RDF data with
    geospatial information.


Getting Started
===============
To get started  with Strabon please have a look at the tutorial for the stRDF
data model and stSPARQL query language, the User Guide, and the Developer Guide.

stRDF and stSPARQL tutorial
	http://www.strabon.di.uoa.gr/files/stSPARQL_tutorial.pdf


stSPARQL Reference
~~~~~~~~~~~~~~~~~~
The reference for the spatial and temporal extension functions defined in 
stSPARQL can be found at http://www.strabon.di.uoa.gr/stSPARQL.


User Guide
~~~~~~~~~~
Assuming that you are familiar with Maven, the following steps need to be
followed in order to use Strabon using Eclipse:

1. Install PostgreSQL from http://www.postgresql.org/download/. At the time of
   this writing the latest PostgreSQL version is 9.1.
2. Install PostGIS from http://postgis.refractions.net/download/. At the time of
   this writing we have tested Strabon with PostGIS 1.5.3. For the
   temporal features of stSPARQL download the PostgreSQL temporal extension
   from here: https://github.com/jeff-davis/PostgreSQL-Temporal and follow
   the installation guide.
3. Install Maven from http://maven.apache.org/download.html. At the time of this
   writing the latest Maven version is 3.0.4. 
4. Install Eclipse from http://www.eclipse.org/downloads/. At the time of this
   writing the latest Eclipse version is 3.7.2.
5. Install the m2e plugin for Eclipse from http://www.eclipse.org/m2e/.
6. Install the MercurialEclipse plugin for Eclipse from
   http://javaforge.com/project/HGE .
7. From Eclipse, go to File --> Import --> Mercurial --> Clone Existing
   Mercurial Repository --> Next. In the URL textarea paste the following
   URL: http://hg.strabon.di.uoa.gr/StrabonUser and then press Next --> Next -->
   Finish. If you used the default settings, you should have a new project named
   StrabonMain. Right click on the project and select Configure --> Convert to
   Maven project. Eclipse will enable Maven dependency management for the
   project, download any dependencies and build the project. 


	Storing stRDF graphs and evaluating stSPARQL queries
	~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
You can see some examples in the classes gr.uoa.di.strabon.example.PostgisExample
and gr.uoa.di.strabon.example.PostgisExample2.

	Tuning PostgreSQL
	~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The default settings of Postgres are rather conservative. As a result, parameter 
tuning is neccessary for speeding up Postgres, therefore Strabon. If you are 
using Strabon to compare its performance against your implementation of 
stSPARQL/GeoSPARQL, you are *strongly* encouraged to contact us using the Strabon
Users mailing list for assistance on tuning Postgres.

You can follow the instructions below for tuning a Postgres server running on an 
Ubuntu machine that is dedicated to PostgreSQL and Strabon.

1. Append the following text at the end of postgresql.conf. 
*Uncomment* the appropriate lines.

### RAM
## 4 GB of RAM
#shared_buffers       =  3GB
#effective_cache_size =  3GB
#maintenance_work_mem =  1GB
#work_mem             =  2GB
## 8 GB of RAM
#shared_buffers       =  5GB
#effective_cache_size =  6GB
#maintenance_work_mem =  2GB
#work_mem             =  5GB
## 16 GB of RAM
#shared_buffers       = 10GB
#effective_cache_size = 14GB
#maintenance_work_mem =  4GB
#work_mem             = 10GB
## 24 GB of RAM
#shared_buffers       = 16GB 
#effective_cache_size = 22GB 
#maintenance_work_mem =  6GB 
#work_mem             = 15GB
## 48 GB of RAM
#shared_buffers       = 32GB
#effective_cache_size = 46GB
#maintenance_work_mem =  8GB
#work_mem             = 30GB
## 64 GB of RAM
# contact us to find out!
### HD
## RAID with ordinary 7.200 disks
#random_page_cost = 3.5 #3.0-3.5
## High-End NAS/SAN
#random_page_cost = 2 #1.5-2.5
## Amazon EBS/Heroku
#random_page_cost = 1.3 #1.1-2.0
## SSD array
#random_page_cost = 2.0 #1.5-2.5
### Planner options
# Increase the following values in order to avoid using the GEQO planner. 
# Small values (<8) reduce planning time but may produce inferior query plans
# 
geqo_threshold = 15 # keep this value larger that the following two parameters
from_collapse_limit = 14
join_collapse_limit = 14
### Misc
default_statistics_target    = 10000 
constraint_exclusion         = on 
checkpoint_completion_target = 0.9 
wal_buffers                  = 32MB 
checkpoint_segments          = 64 
### Connections
max_connections              = 10

2. Append the following lines at the end of /etc/sysctl.conf
*Uncomment* the appropriate lines.

## 4 GB of RAM
#kernel.shmmax = 3758096384
#kernel.shmall = 3758096384
#kernel.shmmni = 4096
## 8 GB of RAM
#kernel.shmmax = 5905580032
#kernel.shmall = 5905580032
#kernel.shmmni = 4096
## 16 GB of RAM
#kernel.shmmax = 11274289152
#kernel.shmall = 11274289152
#kernel.shmmni = 4096
## 24 GB of RAM
#kernel.shmmax = 17716740096
#kernel.shmall = 17716740096
#kernel.shmmni = 4096
## 48 GB of RAM
#kernel.shmmax = 35433480192
#kernel.shmall = 35433480192
#kernel.shmmni = 4224
## 64 GB of RAM
# contact us to find out!

3. Apply all changes by executing

$ sudo sysctl -p
$ sudo /etc/init.d/postgresql restart

4. Prepare for the next run by issuing the command 

$ sudo -u postgres psql -c 'VACUUM ANALYZE;' db

or 

$ psql -c 'VACUUM ANALYZE;' db

where db is the name of the Postgres database that Strabon will use.


Developer Guide
~~~~~~~~~~~~~~~
Assuming that you are familiar with Maven, the following steps need to be
followed in order to use Strabon using Eclipse:

1. Install PostgreSQL from http://www.postgresql.org/download/. At the time of
   this writing the latest PostgreSQL version is 9.1.
2. Install PostGIS from http://postgis.refractions.net/download/. At the time of
   this writing we have tested Strabon with PostGIS 1.5.3. For the
   temporal features of stSPARQL download the PostgreSQL temporal extension
   from here: https://github.com/jeff-davis/PostgreSQL-Temporal and follow
   the installation guide.
3. Install Maven from http://maven.apache.org/download.html. At the time of this
   writing the latest Maven version is 3.0.4. 
4. Install Eclipse from http://www.eclipse.org/downloads/. At the time of this
   writing the latest Eclipse version is 3.7.2.
5. Install the m2e plugin for Eclipse from http://www.eclipse.org/m2e/.
6. Install the MercurialEclipse plugin for Eclipse from
   http://javaforge.com/project/HGE.
7. From Eclipse, go to File --> Import --> Mercurial --> Clone Existing
   Mercurial Repository --> Next. In the URL textarea paste the following
   URL: http://hg.strabon.di.uoa.gr/Strabon and then press Next --> Next -->
   Finish. If you used the default settings, you should have a new project named
   StrabonMain. Right click on the project and select Configure --> Convert to
   Maven project. Eclipse will enable Maven dependency management for the
   project, download any dependencies and build the project. 


Tester Guide
~~~~~~~~~~~~
Assuming again that you are familiar with Maven and Junit these are the steps
you need to follow to test the functionality of Strabon:


* Using Eclipse

	If you want to create a new test:

	1. Import Strabon into Eclipse as explained in the Developer Guide.
	2. Go to strabon-testsuite project.
	3. Create a new folder (Recommended folder name: <test's name>) and place inside the following files:
		3.1. An ntriples or nquads file with the test dataset (with .nt or .nq extension).
		3.2. Pairs of files with sparql test queries and expected test results in xml format.
			 Notice that each pair must have the same name and .rq extension for the queryFile and .srx extension for the resultsFile.
	4. Create a test class that extends TemplateTest class.
	5. If you have followed the recommendations the test is ready. If you have different names or location for your
	files, insert them explicitly in the constructor of the class. WARNING: All prefixes must be placed in file "prefixes" so that
	every time a namespace changes, we have to change it just once.
	
	If you want to run a test:
	
	1. Right-click on the test class.
	2. Select "Run as JUnit Test".
	3. Database properties are retrieved from database.properties file. If you want, you can change a property
	"on the fly" with an environment variable.

	
* Command Line
	
	If you want to run all the tests:
	
	1. Go to Strabon directory (root directory of all the subprojects).
	2. Run "mvn test -DskipTests=false".

ChangeLog
	* Related to bug #58, but it generally enhances Strabon; The runtime
	  component has also been debianized. Dependencies are added to
	  /usr/share/jdeb/lib/strabon-runtime. ClassPath and strabon script
	  have also been updated accordingly (strabon-cmd)

	* Bug #58: Maven creates the debian package conditionally. Use
	  -Ddebian=true in the mvn package command. 

	* Bug #58: Added dependencies - java, postgresql, tomcat. Also, added
	  a strabon-init script that detects the postgis installation path
	  and sets up the template_postgis. Finally the script creates a db
	  called "endpoint" (assuming it does not already exist) that is used
	  for the strabon-endpoint command.

	* Fixed a bug that would return no results when a variable that results 
	  from a BIND clause appears in the FILTER or in the projection of a
	  query and is inside a spatial function.
	  (bug #78: http://bug.strabon.di.uoa.gr/ticket/78)

	* Fixed the precision of the geometry literals in GeoJSON writer. Now,
	  this defaults to the precision model of JTS (equals to 16). Prior to
	  this it was fixed to 4.

	* Now the GeoJSON writer is aware of the kind of typed literals (e.g.,
	  numeric, boolean, calendar, etc.) and prints the corresponding
	  values according to these types. 

	* Fixed a bug that would not return the correct result or would fail 
	  when a spatial function involved a constant and a variable.
	  (bug #74: http://bug.strabon.di.uoa.gr/ticket/74)

	* Fixed a bug that would not return the correct spatial datatype of a
	  projected geometry literal stored in the database.
	  (bug #71: http://bug.strabon.di.uoa.gr/ticket/71)

	* Now, geometries expressed in CRS84 and a `geo:wktLiteral' datatype
	  are assigned the correct CRS, i.e., EPSG:4326 with a longitude
	  latitude axis order.

	* Now, we report a parse error when a constant geometry given in a
	  SELECT or HAVING clause is not well-formed according to the WKT
	  specification.
	  (bug #69: http://bug.strabon.di.uoa.gr/ticket/69)

	* Function `geof:convexHull' is now supported in a Having clause.

	* Fixed older implementation of `strdf:distance' that would expect two
	  arguments instead of three when present in a HAVING clause.

	* Fixed a bug that would cause a class cast exception for grounded
	  expressions with function nesting in the SELECT or HAVING clause.

	* Fixed a bug in which `postgis:ST_Centroid' did not work in the SELECT
	  clause with input geometries from the database.

	* Fixed a bug that `geof:getSRID' would return an integer instead of a
	  the URI corrresponding to the CRS that the input geometry was
	  expressed.
	  (bug #32:  http://bug.strabon.di.uoa.gr/ticket/32)

	* Fixed a bug that would raise an exception for queries involving two
	  UNION graph patterns and a projection to a predicate or a URI in
	  general.
	  (bug #53: http://bug.strabon.di.uoa.gr/ticket/53)

Wed Sep 17 20:14:12 2014   Pyravlos Team

	* Version 3.2.10 released.

	* The test suite of Strabon has now reached a mature level. It
	  contains tests for both stSPARQL and GeoSPARQL query languages and
	  most of the compliance tests covered in the specification of
	  GeoSPARQL (well, only those tests for the components that we have
	  chosen to implement). 

	* The endpoint now parses Accept headers with multiple values and uses
	  the first mimetype corresponding to a valid stSPARQLQueryResultFormat.
	  (bug #64: http://bug.strabon.di.uoa.gr/ticket/64)

	* The endpoint now registers a handler for SIGTERM signals, so that if
	  it is run from the command-line (endpoint-exec), then the closing of
	  the connection will be ultimately called.
	  (bug #40: http://bug.strabon.di.uoa.gr/ticket/40)

	* Strabon goes debian v1. the endpoint-exec jar is now given in a .deb 
          format. By dpkg-ing it, one can run a stand alone strabon endpoint
          locally.
	  (bug #45 http://bug.strabon.di.uoa.gr/ticket/58)

	* Fixed a bug in which Strabon hanged when a unary spatial construct 
	  had a ternary spatial construct as argument. 
	  (bug #45 http://bug.strabon.di.uoa.gr/ticket/45)

	* Fixed a bug according to which spatial functions in ORDER BY were not
	  evaluated at all.
	  (bug #62: http://bug.strabon.di.uoa.gr/ticket/62)

	* Fixed a bug according to which spatial boolean functions appearing
	  in the SELECT clause were raising a NULL pointer exception.
	  (bug #34: http://bug.strabon.di.uoa.gr/ticket/34)

	* Fixed a bug in queries that contain variables that exist inside 
	  spatial functions in select and do not exist in the where clause 
	  made Strabon throw a NULL pointer exception.
	  (bug #51 https://bug.strabon.di.uoa.gr/ticket/51)

	* Various changes in the interface and behavior of Strabon Endpoint:
	  - Revised the logos and the sample queries.
	  - Results query limit is set to zero (unlimited).
	  - Users can now store RDF data that is present in a file on the
	    client side (the file is uploaded to the server).
	  - Credentials are now required for uploading files or executing
	    UPDATE queries, unless Strabon Endpoint runs locally (localhost).
	  - When viewing the results of queries in HTML, any URI is now an
	    active link to itself. Previously, that URI was being DESCRIBEd
	    and one was being transferred to a page viewing the result of the
	    DESCRIBE query.
	  - The DESCRIBE functionality can now be accessed through the menu bar.
	
	* Fixed a bug where strdf:intersection function did not work as an
	  aggregate when used without 'GROUP BY'.
	  (bug #46 https://bug.strabon.di.uoa.gr/ticket/46)


Sat Oct 19 23:37:23 2013   Pyravlos Team

	* Version 3.2.9 released.

	* Added support for the following google chart types: PieChart, 
	AreaChart and ColumnChart to strabon endpoint. 
	
	* Added Timemap as new results format in Strabon endpoint. This 
	option displays results in a map and a timeline on the same time, 
	using the Timemap js library (http://code.google.com/p/timemap/).

examples/teleios/runChain.sh
#! /bin/bash
# 
# This Source Code Form is subject to the terms of the Mozilla Public 
# License, v. 2.0. If a copy of the MPL was not distributed with this file, you
# can obtain one at http://mozilla.org/MPL/2.0/. 
# 
# Copyright (C) 2010, 2011, 2012, 2013 Pyravlos Team 
# 
# http://www.strabon.di.uoa.gr/ 
#
# Author: George Garbis <ggarbis@di.uoa.gr>
# Author: Charalampos (Babis) Nikolaou <charnik@di.uoa.gr>
# Author: Manos Karpathiotakis <mk@di.uoa.gr>
# Author: Konstantina Bereta <Konstantina.Bereta@di.uoa.gr>
#

# If tomcat is standalone then environment variable TOMCATPATH should be set

# Example run command: examples/teleios/runChain.sh -b http://dev.strabon.di.uoa.gr/rdf/data-dump-postgres-9.tgz  -l /home/ggarbis/runChain.log -e http://pathway.di.uoa.gr:8080/endpoint

# Command name
cmd="$(basename ${0})" 
# Get the directory where the script resides
loc="$(cd "$(dirname "${BASH_SOURCE[0]}")" && pwd)"

function help() {                                                               
	echo "Usage: ${cmd} [OPTIONS] "                        
	echo                                                                        
	echo "Execute NOA chain with refinements and measure time."
	echo                                                                        
	echo "OPTIONS can be any of the following"                                  
	echo "  -d,--db			: PostGIS database"                
	echo "  -e,--endpoint			: Strabon Endpoint"
	echo "  -h,--hotposts			: URL where hotspots are stored"
	echo "  -b,--background		: Background data"                                           
	echo "  -l,--log			: Log file"                                           
	echo "  -c,--chain			: Processing chain of hotspots"
	echo "  -p,--persistence		: Value of persistence of discoverFires query"
	echo "  -r,--repeat_in_persistence	: Value of repeat_in_persistence of discoverFires query"     
}

# If no arguments are given it returns miliseconds from 1970-01-01 00:00:00 UTC
# Else if a time (miliseconds form ...) is given it returns the delta between
# given time and current time
function timer()
{
   if [[ $# -eq 0 ]]; then
       t=$(date '+%s%N')
       t=$((t/1000000))
       echo $t
   else
       local  stime=$1
       etime=$(date '+%s%N')
       etime=$((etime/1000000))

       if [[ -z "$stime" ]]; then stime=$etime; fi
       dt=$((etime - stime)) #dt in milliseconds
       dM=$((dt%1000))
       Dt=$((dt/1000)) #delta t in seconds
       ds=$((Dt % 60))
       dm=$(((Dt / 60) % 60))
       dh=$((Dt / 3600))
       printf '%d:%02d:%02d.%03d' $dh $dm $ds $dM
   fi
}

# Handle the postgres service
# -$1: Command for the service
function handlePostgresService()
{
	# find out the postgres service to use
	postgres=$(ls -1 /etc/init.d/| grep postgres | head -1)

	echo "Service ${postgres} received command: $1"
	sudo service ${postgres} $1
}

# Handled a postgres database
# -$1: Command (create/drop/store)
# -$2: Dump file to store (if runscript is given as command)
#	   or 'spatial' to create a spatial database (if create is given as command)
function handlePostgresDatabase() {
	local command=$1
	local db=$2
	shift; shift
	local options="$*"
	case "${command}" in
		create)
            if test "${options}" = "spatial"; then
                options="-T template_postgis"
            elif test ! -z "${options}"; then
				echo "ERROR: only spatial is allowed for create option"
				echo "options: ${options}"
				exit -1
			fi
			echo "Creating database ${db}... with options ${options}"
			createdb -U postgres ${db} ${options}	
			;;
		drop)
			if test ! -z "${options}"; then
				echo "ERROR: dropdb takes no extra options"
				echo "options: ${options}"
				exit -1
			fi
			echo "Dropping database ${db}..."
			dropdb -U postgres ${db} 
			;;
		vacuum)
			if test "${options}" = "analyze"; then
				psql -U postgres ${db} -c 'VACUUM ANALYZE' 
                echo "VACUUM ANALYZE ${db}"
			else
				psql -U postgres ${db} -c 'VACUUM' 
                echo "VACUUM ${db}"
			fi
			;;
		runscript)
			if test ! -f "${options}"; then
				echo "ERROR: No dump file to run"
				exit -1
			fi
			echo "Storing dump file ${options} in database ${db}..."
			psql -U postgres ${db} -f ${options}
			;;
	esac
}

# Handle the tomcat service
# -$1: Command for the service
function handleTomcatService()
{
	if test ! -z "${TOMCATPATH}" ; then
		case "${1}" in
			start)
				${TOMCATPATH}/bin/startup.sh
				;;
			stop)
				${TOMCATPATH}/bin/shutdown.sh
				;;
			restart)
				${TOMCATPATH}/bin/startup.sh
				${TOMCATPATH}/bin/shutdown.sh
				;;	
		esac	
		return
	# find out the tomcat service to use
	elif test -s /etc/fedora-release ; then
		tomcat="tomcat"
	#elif test -s /etc/centos-release ; then
	#elif test -s /etc/yellowdog-release ; then
	#elif test -s /etc/redhat-release ; then
	#elif test -s /etc/SuSE-release ; then
	#elif test -s /etc/gentoo-release ; then
	elif test -s /etc/lsb-release ; then # Ubuntu
		tomcat=$(ls -1 /etc/init.d/| grep tomcat | head -1)
	elif test -s /etc/debian_version ; then
		tomcat="tomcat"
	fi

	# check for service availability
	if ! test -e "/etc/init.d/${tomcat}"; then
		echo "ERROR: No tomcat service found"
		exit -1
	fi

	echo "Service ${tomcat} received command: $1"
	sudo service ${tomcat} $1
}


# get the main version of postgres
function getPostgresMainVersion() {
	echo $(sudo service ${postgres} status | grep -o '.\..' | cut -b 1)
}

# It stores the backgroud data
# - $1: database
# - $2: backgound data file
function storeBackgroundData() {
	local db=$1
	local bgFile=$2

	if test -f ${bgFile}; then
		handlePostgresDatabase runscript ${db} ${bgFile}	
	elif test "${bgFile:0:7}" = "http://"; then
		curl -s ${bgFile} | tar xzf - -O > /tmp/bgFiles$$.sql
#		wget ${bgFile} -O /tmp/bgFile$$.tar.gz
#		tar xzf /tmp/bgFile$$.tar.gz		
		handlePostgresDatabase runscript ${db} /tmp/bgFiles$$.sql
#		rm /tmp/bgFile$$.tar.gz
		rm /tmp/bgFiles$$.sql
	else
		echo "Backgound file not foung"
		exit -1
	fi
	handlePostgresDatabase vacuum ${db} analyze
}

# Handle Stabon Endpoint
# - $1: endpoint
# - $2: command (store/query)
# - $2: file/query
function handleStrabonEndpoint(){
	endpoint=$1
	command=$2
	options=$3

	endpointScript=${loc}/../../scripts/endpoint
	case ${command} in
		store)
			url=${options}

			tmr1=$(timer)
			#${endpointScript} store ${endpoint} N-Triples -u ${url}
			#read t
			${endpointScript} store ${endpoint} N-Triples -u ${url}
			tmr2=$(timer)

    		# execute an explicit VACUUM ANALYZE when a query takes longer than it should
    		duration=$((tmr2-tmr1))
    		if test ${duration} -ge 30000; then
                handlePostgresDatabase vacuum ${db} analyze
#    			psql -U postgres ${DB} -c 'VACUUM ANALYZE' 
#    			echo "Explicit VACUUM ANALYZE"
        	    tmr2=$(timer)
    		fi
			printf '%s ' $((tmr2-tmr1)) >> ${logFile}
			;;
		query)
			query=${options}
			tmr1=$(timer)
			#${endpointScript} query ${endpoint} "${query}"
			#read t
			${endpointScript} query ${endpoint} "${query}"
			tmr2=$(timer)
			printf '%s ' $((tmr2-tmr1)) >> ${logFile} 
			;;
		update)
			update=${options}
			tmr1=$(timer)
			#${endpointScript} update ${endpoint} "${update}"
			#read t
			${endpointScript} update ${endpoint} "${update}"
			tmr2=$(timer)
			printf '%s ' $((tmr2-tmr1)) >> ${logFile}
			;;
		*)
			echo "ERROR: Unknown endpoint command"
			exit -1
			;;
	esac
}

# default values
endpoint="http://teleios3.di.uoa.gr:8080/endpoint"
db="NOA2012"
hotspotsURL="http://jose.di.uoa.gr/rdf/hotspots/MSG1"
#                                 ./examples/teleios/data/data-dump-9.sql
bgFile="http://dev.strabon.di.uoa.gr/rdf/data-dump-9.sql"
logFile="/home/ggarbis/runChain.log"

chain="DynamicThresholds"
persistence=10
repeatInPers=3

# read script options
while test $# -gt 0 -a "X${1:0:1}" == "X-"; do
    case "${1}" in
        --help)
            help
            exit 0
            ;;
        -e|--endpoint)
            shift
			endpoint=${1}
            shift
            ;;
        -d|--db)
            shift
			db=${1}
			shift
			;;
		-h|--hotspots)
            shift
			hotspots_url=${1}
			shift
			;;
		-b|--background)
            shift
			bgFile=${1}
			shift
			;;
		-l|--log)
            shift
			logFile=${1}
			shift
			;;
        -c|--chain)
            shift
            chain=${1}
            shift
            ;;
        -p|--persistence)
            shift
            persistence=${1}
            shift
            ;;
        -r|--repeat_in_persistence)
            shift
            repeat_in_persistence=${1}
            shift
            ;;
		*)
			echo "unknown argument ${1}"
			help
			exit -1
			;;
	esac
done

echo "endpoint: ${endpoint}"
echo "db: ${db}"
echo "hotspots: ${hotspotsURL}"
echo "background: ${bgFile}"
echo "logFile: ${logFile}"

echo > ${logFile}
echo 2222
instantiate=${loc}/instantiate.sh

#Initialize (stop tomcat, restart postgres, drop/create database, start tomcat)
echo 111
handleTomcatService stop
handlePostgresService restart

handlePostgresDatabase drop ${db}
handlePostgresDatabase create ${db}

storeBackgroundData ${db} ${bgFile} # ~/Temp/Kallikratis-Coastline-Corine-postgres-9.sql

handleTomcatService start
# Wait until tomcat server is up
until [ "`curl --silent --show-error --connect-timeout 1 -I ${endpoint} | grep 'Coyote'`" != "" ]; do sleep 1; done;


#${loc}/../../scripts/endpoint query ${endpoint} size 
#exit -1
echo "Timestamp Store Municipalities DeleteInSea InvalidForFires RefineInCoast TimePersistence DiscoverHotspots" > ${logFile}
echo > /home/ggarbis/discoverFires.log
echo > /home/ggarbis/discover.log

years="2012" #"2007 2008 2010 2011"
for y in ${years}; do
#    hotspots="`ls /var/www/hotspots/${y} | sort | grep -o 'HMSG.*\.nt'`"
	# get hotpost URLS
	for hot in $(curl -s ${hotspotsURL}/${y}/ | grep -o '>HMSG.*\.nt' | colrm 1 1); do
#	for hot in ${hotspots}; do
		file="${hotspotsURL}/${y}/${hot}"

		time_status=$(echo ${hot} | egrep -o '[[:digit:]]{6}_[[:digit:]]{4}')
		
		# get sensor
		sensor=$(echo ${hot} | grep -o 'MSG.')

		# get satellite and set number of acquisitions per hour
		if test "${sensor}" = "MSG1"; then
			sensor="MSG1_RSS"
		fi

		# get time information for acquisition and construct timestamp
		year="20$(expr substr ${time_status} 1 2)"
		month=$(expr substr ${time_status} 3 2)
		day=$(expr substr ${time_status} 5 2)
		time2=$(expr substr ${time_status} 8 2)
		time2="${time2}:$(expr substr ${time_status} 10 2)"

		# construct timestamp
		timestamp="${year}-${month}-${day}T${time2}:00"
		# print timestamp
		echo -n "${timestamp} " >> ${logFile}
        
		handleStrabonEndpoint ${endpoint} store ${file}
        echo "Processing File ${file}" ; # read t

		# Insert Municipalities	
		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} ${loc}/insertMunicipalities.rq`"
#        echo "Insert Municipalities: ${update}" ; read t
        handleStrabonEndpoint ${endpoint} update "${update}"


		# Delete Sea Hotspots
		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} ${loc}/deleteSeaHotspots.rq`"
#       echo "Delete Sea Hotspots: ${update}" ; read t
		handleStrabonEndpoint ${endpoint} update "${update}"

		# Invalid For Fires
		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} ${loc}/landUseInvalidForFires.rq`"
#       echo "Invalid For Fires: ${update}" ; read t
		handleStrabonEndpoint ${endpoint} update "${update}"
 
#        # Delete Reflections
#    	minTime=`date --date="${year}-${month}-${day} ${time2}:00 EEST -60 minutes" +%Y-%m-%dT%H:%M:00`
#		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} -m ${minTime} ${loc}/deleteReflections.rq`"
##        echo "Delete Reflections: ${update}" ;
#		handleStrabonEndpoint ${endpoint} update "${update}"

		# Refine Partial Sea Hotspots
		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} ${loc}/refinePartialSeaHotspots.rq`"
#       echo "Refine Partial Sea Hotspots: ${update}" ; read t
		handleStrabonEndpoint ${endpoint} update "${update}"

		# Refine Time Persistence
		minTime=`date --date="${year}-${month}-${day} ${time2}:00 EEST -30 minutes" +%Y-%m-%dT%H:%M:00`
		update="`${instantiate} -t ${timestamp} -c ${chain} -s ${sensor} -m ${minTime} ${loc}/refineTimePersistence.rq`"
#       echo "Refine Time Persistence: ${update}" ; read t
		handleStrabonEndpoint ${endpoint} update "${update}" #2>&1 | tee /home/ggarbis/timePersistence.log

		#psql -U postgres -d ${DB} -c 'VACUUM ANALYZE;';
        
		# Discover
		minTime=`date --date="${year}-${month}-${day} 00:00 EEST" +%Y-%m-%dT%H:%M:00`
		maxTime=`date --date="${year}-${month}-${day} 23:59 EEST" +%Y-%m-%dT%H:%M:00`
        query="`${instantiate} -c ${chain} -s ${sensor} -m ${minTime} -M ${maxTime} ${loc}/discover.rq`"
#        echo "Discover: ${query}" ; #read t
		handleStrabonEndpoint ${endpoint} query "${query}" &>> /home/ggarbis/discover.log
#    
#		# Discover Fires
#		minTime=`date --date="${year}-${month}-${day} 00:00 EEST" +%Y-%m-%dT%H:%M:00`
#		maxTime=`date --date="${year}-${month}-${day} 23:59 EEST" +%Y-%m-%dT%H:%M:00`
#        query="`${instantiate} -c ${chain} -s ${sensor} -m ${minTime} -M ${maxTime} -p 10 -r 3 ${loc}/discoverFires.rq`"
##        echo "Discover Fires: ${query}" ; #read t
#		handleStrabonEndpoint ${endpoint} query "${query}" &>> /home/ggarbis/discoverFires.log

		# Add a new line
        echo >> ${logFile}    
	done
done
